src/Raven.Server/Documents/Queries/QueryMetadataCache.cs

```diff
@@ -0,0 +1,67 @@
+﻿using Raven.Client.Documents.Queries;
+using Raven.Server.Documents.Queries.Parser;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Queries
+{
+    public static class QueryMetadataCache
+    {
+        private const int CacheSize = 512;
+
+        private static readonly QueryMetadata[] Cache = new QueryMetadata[CacheSize];
+
+        public static bool TryGetMetadata(IndexQueryBase<BlittableJsonReaderObject> query, JsonOperationContext context, out ulong metadataHash, out QueryMetadata metadata)
+        {
+            metadataHash = 0;
+            metadata = null;
+
+            if (query == null || query.Query == null || query.QueryParameters == null || query.QueryParameters.Count == 0)
+                return false;
+
+            if (TryGetQueryMetadataHash(query, context, out metadataHash) == false)
+                return false;
+
+            metadata = Cache[metadataHash % CacheSize];
+            if (metadata == null || metadata.CacheKey != metadataHash)
+                return false;
+
+            return true;
+        }
+
+        public static void MaybeAddToCache(QueryMetadata metadata, string indexName)
+        {
+            if (metadata.CanCache == false)
+                return;
+
+            if (metadata.IsDynamic)
+                metadata.DynamicIndexName = indexName;
+
+            Cache[metadata.CacheKey % CacheSize] = metadata;
```

Comment:
If there are two threads trying to set the same location at once.
Or if there is a hash collision. That is actually fine, because we'll get last write wins, but two queries that have the same hash will always kick one another off the cache.